kclvm/sema/src/resolver/doc.rs

```diff
@@ -0,0 +1,599 @@
+use regex::Regex;
+use std::collections::HashSet;
+use std::fs::File;
+use std::io::prelude::*;
+use std::iter::Iterator;
+use std::path::PathBuf;
+
+// strip leading and trailing triple quotes in the original docstring content
+fn strip_quotes(original: &mut String) {
+    let quote = original.chars().next().unwrap();
+    let pattern = format!("(?s)^{char}{{3}}(.*?){char}{{3}}$", char = quote);
+    let re = Regex::new(&pattern).unwrap();
+    let caps = re.captures(&original);
+    let result = match caps {
+        Some(caps) => caps,
+        None => return,
+    };
+    let content = result[1].to_owned();
+    *original = content;
+}
+
+fn expand_tabs(s: &str, spaces_per_tab: usize) -> String {
+    s.replace("\t", &" ".repeat(spaces_per_tab))
+}
+
+// Clean up indentation by removing any common leading whitespace
+// on all lines after the first line.
+pub fn clean_doc(doc: &mut String) {
+    let tab_expanded = expand_tabs(&doc, 4);
+    let mut lines: Vec<&str> = tab_expanded.split('\n').collect();
```

Comment:
Considering the windows platform newline `\r\n`.

Reply:
> Considering the windows platform newline `\r\n`.

the docstring here is not the exact original docstring read from the KCL file: the `\r\n` was already processed before kclvm parser phase